jerry-core/ecma/builtin-objects/ecma-builtin-date.c

```diff
@@ -725,53 +726,74 @@ ecma_builtin_date_dispatch_construct (const ecma_value_t *arguments_list_p, /**<
 
   ecma_number_t prim_value_num = ECMA_NUMBER_ZERO;
 
+  /* 20.4.2.3 */
   if (arguments_list_len == 0)
   {
     prim_value_num = ecma_builtin_date_now_helper ();
   }
+  /* 20.4.2.2 */
   else if (arguments_list_len == 1)
   {
-    ecma_value_t prim_comp_value = ecma_op_to_primitive (arguments_list_p[0], ECMA_PREFERRED_TYPE_NUMBER);
+    ecma_value_t argument = arguments_list_p[0];
+    ecma_object_t *arg_obj = NULL;
 
-    if (ECMA_IS_VALUE_ERROR (prim_comp_value))
+    /* 4.a */
+    if (ecma_is_value_object (argument))
     {
-      ecma_deref_object (obj_p);
-      return prim_comp_value;
+      arg_obj = ecma_get_object_from_value (argument);
     }
 
-    if (ecma_is_value_string (prim_comp_value))
+    if (arg_obj && ecma_object_class_is (arg_obj, LIT_MAGIC_STRING_DATE_UL))
+    {
+      ecma_extended_object_t *arg_ext_object_p = (ecma_extended_object_t *) arg_obj;
+      prim_value_num = *ECMA_GET_INTERNAL_VALUE_POINTER (ecma_number_t, arg_ext_object_p->u.class_prop.u.value);
+    }
+    /* 4.b */
+    else
     {
-      ecma_value_t parse_res_value = ecma_builtin_date_parse (ecma_make_object_value (obj_p), prim_comp_value);
+      ecma_value_t prim_comp_value = ecma_op_to_primitive (argument, ECMA_PREFERRED_TYPE_NUMBER);
 
-      if (ECMA_IS_VALUE_ERROR (parse_res_value))
+      if (ECMA_IS_VALUE_ERROR (prim_comp_value))
       {
         ecma_deref_object (obj_p);
-        ecma_free_value (prim_comp_value);
-        return parse_res_value;
+        return prim_comp_value;
       }
 
-      prim_value_num = ecma_get_number_from_value (parse_res_value);
+      if (ecma_is_value_string (prim_comp_value))
+      {
+        ecma_value_t parse_res_value = ecma_builtin_date_parse (ecma_make_object_value (obj_p), prim_comp_value);
 
-      ecma_free_value (parse_res_value);
-    }
-    else
-    {
-      ecma_value_t prim_value = ecma_op_to_number (arguments_list_p[0]);
+        if (ECMA_IS_VALUE_ERROR (parse_res_value))
+        {
+          ecma_deref_object (obj_p);
+          ecma_free_value (prim_comp_value);
+          return parse_res_value;
+        }
 
-      if (ECMA_IS_VALUE_ERROR (prim_value))
-      {
-        ecma_deref_object (obj_p);
-        ecma_free_value (prim_comp_value);
-        return prim_value;
+        prim_value_num = ecma_get_number_from_value (parse_res_value);
+
+        ecma_free_value (parse_res_value);
       }
+      else
+      {
+        ecma_value_t prim_value = ecma_op_to_number (argument);
```

Comment:
I'd rather use `ecma_get_number` here.

Reply:
I haven't modified this code, just added to whitespaces. I prefer this optimization in a separated PR, because it isn't related to this fix.

Reply:
Fine!